rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/custom_service_info.hpp

```diff
@@ -26,125 +26,28 @@
 #include "fastrtps/subscriber/Subscriber.h"
 #include "fastrtps/subscriber/SubscriberListener.h"
 #include "fastrtps/subscriber/SampleInfo.h"
+#include "fastrtps/TopicDataType.h"
 
-class ServiceListener;
+#include "rmw_fastrtps_shared_cpp/custom_service_info.hpp"
 
-typedef struct CustomServiceInfo
+struct CustomServiceInfoImpl : public CustomServiceInfo
 {
-  void * request_type_support_;
-  void * response_type_support_;
-  eprosima::fastrtps::Subscriber * request_subscriber_;
-  eprosima::fastrtps::Publisher * response_publisher_;
-  ServiceListener * listener_;
-  eprosima::fastrtps::Participant * participant_;
-  const char * typesupport_identifier_;
-} CustomServiceInfo;
-
-typedef struct CustomServiceRequest
-{
-  eprosima::fastrtps::rtps::SampleIdentity sample_identity_;
-  eprosima::fastcdr::FastBuffer * buffer_;
-
-  CustomServiceRequest()
-  : buffer_(nullptr) {}
-} CustomServiceRequest;
-
-class ServiceListener : public eprosima::fastrtps::SubscriberListener
-{
-public:
-  explicit ServiceListener(CustomServiceInfo * info)
-  : info_(info), list_has_data_(false),
-    conditionMutex_(nullptr), conditionVariable_(nullptr)
+  inline bool take(void * ros_request,
```

Comment:
It would be great if at least all classes and public functions could be documented.